src/frontend/src/optimizer/plan_node/logical_recursive_union.rs

```diff
@@ -0,0 +1,124 @@
+use risingwave_common::bail_not_implemented;
+use risingwave_common::util::column_index_mapping::ColIndexMapping;
+use smallvec::{smallvec, SmallVec};
+
+use super::expr_visitable::ExprVisitable;
+use super::generic::GenericPlanRef;
+use super::utils::{childless_record, Distill};
+use super::{
+    generic, ColPrunable, ColumnPruningContext, ExprRewritable, Logical, PlanBase, PlanTreeNode,
+    PredicatePushdown, PredicatePushdownContext, RewriteStreamContext, ToBatch, ToStream,
+    ToStreamContext,
+};
+use crate::binder::ShareId;
+use crate::error::Result;
+use crate::utils::Condition;
+use crate::PlanRef;
+
+/// `LogicalRecursiveUnion` returns the union of the rows of its inputs.
+/// note: if `all` is false, it needs to eliminate duplicates.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct LogicalRecursiveUnion {
+    pub base: PlanBase<Logical>,
+    core: generic::RecursiveUnion<PlanRef>,
+}
+
+impl LogicalRecursiveUnion {
+    pub fn new(base_plan: PlanRef, recursive: PlanRef, id: ShareId) -> Self {
+        let core = generic::RecursiveUnion {
+            base: base_plan,
+            recursive,
+            id,
+        };
+        let base = PlanBase::new_logical_with_core(&core);
+        LogicalRecursiveUnion { base, core }
+    }
+
+    pub fn create(base_plan: PlanRef, recursive: PlanRef, id: ShareId) -> PlanRef {
+        Self::new(base_plan, recursive, id).into()
+    }
+
+    pub(super) fn pretty_fields(base: impl GenericPlanRef, name: &str) -> XmlNode<'_> {
+        childless_record(name, vec![("id", Pretty::debug(&base.id().0))])
+    }
+}
+
+impl PlanTreeNode for LogicalRecursiveUnion {
+    fn inputs(&self) -> SmallVec<[PlanRef; 2]> {
+        smallvec![self.core.base.clone(), self.core.recursive.clone()]
+    }
+
+    fn clone_with_inputs(&self, inputs: &[PlanRef]) -> PlanRef {
+        let mut inputs = inputs.iter().cloned();
+        Self::create(inputs.next().unwrap(), inputs.next().unwrap(), self.core.id)
+    }
+}
+
+impl Distill for LogicalRecursiveUnion {
+    fn distill<'a>(&self) -> XmlNode<'a> {
+        Self::pretty_fields(&self.base, "LogicalRecursiveUnion")
+    }
+}
+
+impl ColPrunable for LogicalRecursiveUnion {
+    fn prune_col(&self, required_cols: &[usize], ctx: &mut ColumnPruningContext) -> PlanRef {
+        let new_inputs = self
+            .inputs()
+            .iter()
+            .map(|input| input.prune_col(required_cols, ctx))
+            .collect_vec();
+        let new_plan = self.clone_with_inputs(&new_inputs);
+        self.ctx()
+            .insert_rcte_cache_plan(self.core.id, new_plan.clone());
+        new_plan
+    }
+}
+
+impl ExprRewritable for LogicalRecursiveUnion {}
+
+impl ExprVisitable for LogicalRecursiveUnion {}
+
+impl PredicatePushdown for LogicalRecursiveUnion {
+    fn predicate_pushdown(
+        &self,
+        _predicate: Condition,
+        _ctx: &mut PredicatePushdownContext,
+    ) -> PlanRef {
+        self.clone().into()
```

Comment:
yep this should be handled once the optimization is enabled - since there is no actual implementation for predicate pushdown at present, it's alright.

Reply:
I think we need to change it to `unimplemented!()`, if we don't handle it right now, because the current implementation doesn't handle the push down predicate and it will generate a wrong plan.

Reply:
> it will generate a wrong plan

I don't quite get this, why directly cloning the current plan without predicate pushdown will lead to a wrong plan? 👀

Reply:
If you don't want to push down the predicate you should implement it like this, because the predicate in this function couldn't be thrown away, at least we need to generate a filter for it.
```
fn predicate_pushdown(
        &self,
        predicate: Condition,
        ctx: &mut PredicatePushdownContext,
    ) -> PlanRef {
        // No pushdown.
        gen_filter_and_pushdown(self, predicate, Condition::true_cond(), ctx)
    }
```

Reply:
update the implementation for `predicate_pushdown`, I think it's reasonable to push predicate down to the `base` and `recursive` plan.